### PR TITLE
Library cleanup and refactor

### PR DESCRIPTION
**Pull request by mbolivar**
This is the first step towards breaking our codebase up into something that other applications can build on top of.

- Start with some cleanup, like replacing the nonstandard logging facilities with Zephyr's system logs
- Separate the code that's just our application from the core "FOTA" and other library code.

Next steps:

- tcp.c in general is an obstacle to making a clean separation. Once the gateway is stable, I think it'd be good to refactor these to remove the TCP interface lock, separate the Bluemix and hawkBit network resources, let them do I/O in parallel, and debug the issues.
- Where to move this code so other apps can use it. My vote would be into $ZEPHYR_BASE, e.g. zephyr/lib/ or zephyr/ext/. Other ideas add build system complexity, which may be wasted work depending on Zephyr's build system future.
- Add additional Kconfig logic to e.g. make Bluemix default to off, etc.

### CODE REVIEW

**Comment by mbolivar** [comment]
I'll fix the checkpatch issues. The build errors are because it needs some Zephyr patches, which I'll post separately.

File: src/hawkbit.c [hunk 1]
```diff
@@ -196,6 +196,93 @@ static int handle_headers_complete_download(struct http_parser *parser)
...
+}
+
+/**
+ * @brief Update an ACID of a given type on flash.
+ *
+ * @param type ACID type to update
+ * @param acid New ACID value
+ * @return 0 on success, negative on error.
+ */
+static int hawkbit_acid_update(hawkbit_acid_t type, uint32_t acid)
```
Comment (mike-scott) [inline src/hawkbit.c]: Function vs. variable vs enumeration naming is really confusing here: we have functions like "hawkbit_acid_update" and then "hawkbit_update_acid" below it.

File: src/hawkbit.c [hunk 2]
```diff
@@ -605,8 +692,8 @@ int hawkbit_ddi_poll(void)
 	jsmntok_t jtks[60];	/* Enough for one artifact per SM */
 	int i, ret, len, ntk;
 	static hawkbit_update_action_t hawkbit_update_action;
-	static int hawkbit_acid = 0;
```
Comment (mike-scott) [inline src/hawkbit.c]: Continuing with the name confusion:
We're changing from hawkbit_acid as the parsed value from the server's JSON to something more normal like "json_acid" (good change) and then renaming what was boot_acid (which didn't mean much to begin with) to another variable which now means even less "hawkbit_acid".  The hawkbit_acid structure is specific to device storage.  Perhaps we can name it something like "device_storage_acid"?

File: src/hawkbit.h [hunk 3]
```diff
@@ -13,6 +13,18 @@
 
 extern int poll_sleep;
 
+typedef enum {
+	HAWKBIT_ACID_CURRENT = 0,
+	HAWKBIT_ACID_UPDATE,
+} hawkbit_acid_t;
```
Comment (mike-scott) [inline src/hawkbit.h]: Maybe call this hawkbit_device_acid_t or hawkbit_device_storage_acid_t?

File: src/hawkbit.h [hunk 4]
```diff
@@ -13,6 +13,18 @@
...
+
+struct hawkbit_acid {
```
Comment (mike-scott) [inline src/hawkbit.h]: And rename this to hawkbit_device_acid or hawkbit_device_storage_acid?

File: src/hawkbit.h [hunk 5]
```diff
@@ -13,6 +13,18 @@
...
+	uint32_t current;
+	uint32_t update;
+};
+
+int hawkbit_init(void);
+void hawkbit_acid_read(struct hawkbit_acid *hawkbit_acid);
```
Comment (mike-scott) [inline src/hawkbit.h]: hawkbit_device_acid_read?

**Comment by mbolivar** [comment]
Now updated. Here's the diff from v1:

```
diff --git a/src/lib/fota_shell.c b/src/lib/fota_shell.c
index daf88c5..30cab16 100644
--- a/src/lib/fota_shell.c
+++ b/src/lib/fota_shell.c
@@ -24,7 +24,7 @@
 
 static int fota_cmd_get_acid(int argc, char *argv[])
 {
-	struct hawkbit_acid acid;
+	struct hawkbit_dev_storage_acid acid;
 
 	if (argc != 1) {
 		printk("\"get acid\" takes no arguments.\n");
@@ -33,7 +33,7 @@ static int fota_cmd_get_acid(int argc, char *argv[])
 
 	ARG_UNUSED(argv);
 
-	hawkbit_acid_read(&acid);
+	hawkbit_device_acid_read(&acid);
 	printk("ACID: current=%d, update=%d\n", acid.current, acid.update);
 	return 0;
 }
diff --git a/src/lib/hawkbit.c b/src/lib/hawkbit.c
index 21e5e84..f6f5a59 100644
--- a/src/lib/hawkbit.c
+++ b/src/lib/hawkbit.c
@@ -84,6 +84,11 @@ typedef enum {
 	HAWKBIT_EXEC_RESUMED,
 } hawkbit_exec_status_t;
 
+typedef enum {
+	HAWKBIT_ACID_CURRENT = 0,
+	HAWKBIT_ACID_UPDATE,
+} hawkbit_dev_acid_t;
+
 #define HAWKBIT_RX_TIMEOUT	K_SECONDS(3)
 
 /* Utils */
@@ -195,10 +200,10 @@ static int handle_headers_complete_download(struct http_parser *parser)
 	return 1;
 }
 
-void hawkbit_acid_read(struct hawkbit_acid *hawkbit_acid)
+void hawkbit_device_acid_read(struct hawkbit_device_acid *device_acid)
 {
-	flash_read(flash_dev, FLASH_STATE_OFFSET, hawkbit_acid,
-					sizeof(*hawkbit_acid));
+	flash_read(flash_dev, FLASH_STATE_OFFSET, device_acid,
+		   sizeof(*device_acid));
 }
 
 /**
@@ -208,17 +213,18 @@ void hawkbit_acid_read(struct hawkbit_acid *hawkbit_acid)
  * @param acid New ACID value
  * @return 0 on success, negative on error.
  */
-static int hawkbit_acid_update(hawkbit_acid_t type, uint32_t acid)
+static int hawkbit_device_acid_update(hawkbit_dev_acid_t type,
+				      uint32_t new_value)
 {
-	struct hawkbit_acid hawkbit_acid;
+	struct hawkbit_device_acid device_acid;
 	int ret;
 
-	flash_read(flash_dev, FLASH_STATE_OFFSET, &hawkbit_acid,
-					sizeof(hawkbit_acid));
+	flash_read(flash_dev, FLASH_STATE_OFFSET, &device_acid,
+		   sizeof(device_acid));
 	if (type == HAWKBIT_ACID_UPDATE) {
-		hawkbit_acid.update = acid;
+		device_acid.update = new_value;
 	} else {
-		hawkbit_acid.current = acid;
+		device_acid.current = new_value;
 	}
 
 	flash_write_protection_set(flash_dev, false);
@@ -229,37 +235,22 @@ static int hawkbit_acid_update(hawkbit_acid_t type, uint32_t acid)
 	}
 
 	flash_write_protection_set(flash_dev, false);
-	ret = flash_write(flash_dev, FLASH_STATE_OFFSET, &hawkbit_acid,
-			  sizeof(hawkbit_acid));
+	ret = flash_write(flash_dev, FLASH_STATE_OFFSET, &device_acid,
+			  sizeof(device_acid));
 	flash_write_protection_set(flash_dev, true);
 	return ret;
 }
 
-static int hawkbit_update_acid(struct hawkbit_acid *acid)
-{
-	int ret = 0;
-	if (acid->update != -1) {
-		ret = hawkbit_acid_update(HAWKBIT_ACID_CURRENT, acid->update);
-		if (!ret) {
-			hawkbit_acid_read(acid);
-			SYS_LOG_INF("ACID updated, current %d, update %d",
-				    acid->current, acid->update);
-		} else {
-			SYS_LOG_ERR("Failed to update ACID: %d", ret);
-		}
-	}
-	return ret;
-}
-
 int hawkbit_init(void)
 {
-	int ret;
-	struct hawkbit_acid acid;
+	int ret = 0;
+	struct hawkbit_device_acid init_acid;
 	uint8_t boot_status;
 
 	/* Update boot status and acid */
-	hawkbit_acid_read(&acid);
-	SYS_LOG_INF("ACID: current %d, update %d", acid.current, acid.update);
+	hawkbit_device_acid_read(&init_acid);
+	SYS_LOG_INF("ACID: current %d, update %d",
+		    init_acid.current, init_acid.update);
 	boot_status = boot_status_read();
 	SYS_LOG_INF("Current boot status %x", boot_status);
 	if (boot_status == BOOT_STATUS_ONGOING) {
@@ -274,12 +265,19 @@ int hawkbit_init(void)
 			SYS_LOG_DBG("Erased flash bank at offset %x",
 				    FLASH_BANK1_OFFSET);
 		}
-		ret = hawkbit_update_acid(&acid);
-		if (ret) {
-			return ret;
+		if (init_acid.update != -1) {
+			ret = hawkbit_device_acid_update(HAWKBIT_ACID_CURRENT,
+						  init_acid.update);
+		}
+		if (!ret) {
+			hawkbit_device_acid_read(&init_acid);
+			SYS_LOG_INF("ACID updated, current %d, update %d",
+				    init_acid.current, init_acid.update);
+		} else {
+			SYS_LOG_ERR("Failed to update ACID: %d", ret);
 		}
 	}
-	return 0;
+	return ret;
 }
 
 static void hawkbit_header_cb(struct net_context *context,
@@ -692,8 +690,8 @@ int hawkbit_ddi_poll(void)
 	jsmntok_t jtks[60];	/* Enough for one artifact per SM */
 	int i, ret, len, ntk;
 	static hawkbit_update_action_t hawkbit_update_action;
-	static int json_acid = 0;
-	struct hawkbit_acid hawkbit_acid;
+	static int json_acid;
+	struct hawkbit_device_acid device_acid;
 	struct json_data_t json = { NULL, 0 };
 	char deployment_base[40];	/* TODO: Find a better value */
 	char download_http[200];	/* TODO: Find a better value */
@@ -892,16 +890,16 @@ int hawkbit_ddi_poll(void)
 		}
 	}
 
-	hawkbit_acid_read(&hawkbit_acid);
+	hawkbit_device_acid_read(&device_acid);
 
-	if (hawkbit_acid.current == json_acid) {
+	if (device_acid.current == json_acid) {
 		/* We are coming from a successful flash, update the server */
 		hawkbit_report_update_status(json_acid,
 					     tcp_buf, TCP_RECV_BUF_SIZE,
 					     HAWKBIT_RESULT_SUCCESS,
 					     HAWKBIT_EXEC_CLOSED);
 		return 0;
-	} else if (hawkbit_acid.update == json_acid) {
+	} else if (device_acid.update == json_acid) {
 		/* There was already an atempt, so announce a failure */
 		hawkbit_report_update_status(json_acid,
 					     tcp_buf, TCP_RECV_BUF_SIZE,
@@ -948,7 +946,7 @@ int hawkbit_ddi_poll(void)
 
 	SYS_LOG_INF("Triggering OTA update.");
 	boot_trigger_ota();
-	ret = hawkbit_acid_update(HAWKBIT_ACID_UPDATE, json_acid);
+	ret = hawkbit_device_acid_update(HAWKBIT_ACID_UPDATE, json_acid);
 	if (ret != 0) {
 		SYS_LOG_ERR("Failed to update ACID: %d", ret);
 		return -1;
diff --git a/src/lib/hawkbit.h b/src/lib/hawkbit.h
index 7d90d79..6279cf4 100644
--- a/src/lib/hawkbit.h
+++ b/src/lib/hawkbit.h
@@ -13,18 +13,13 @@
 
 extern int poll_sleep;
 
-typedef enum {
-	HAWKBIT_ACID_CURRENT = 0,
-	HAWKBIT_ACID_UPDATE,
-} hawkbit_acid_t;
-
-struct hawkbit_acid {
+struct hawkbit_device_acid {
 	uint32_t current;
 	uint32_t update;
 };
 
 int hawkbit_init(void);
-void hawkbit_acid_read(struct hawkbit_acid *hawkbit_acid);
+void hawkbit_device_acid_read(struct hawkbit_device_acid *device_acid);
 int hawkbit_ddi_poll(void);
 
 #endif	/* __FOTA_HAWKBIT_H__ */
```

**Comment by mike-scott** [comment]
LGTM.  Let's get @rsalveti opinion.

**Comment by mbolivar** [comment]
Updated to add a config option to control the library code. v2-v3 diff:

```
diff --git a/Kconfig b/Kconfig
index 313e6ea..7cbb7ba 100644
--- a/Kconfig
+++ b/Kconfig
@@ -53,6 +53,22 @@ config FOTA_DEVICE_SOC_SERIES_STM32F4X
 	select TEST_RANDOM_GENERATOR
 	default n
 
+config SYS_LOG_FOTA_LEVEL
+	int "FOTA log level"
+	depends on SYS_LOG && FOTA_DEVICE
+	default 0
+	range 0 4
+	help
+	  Set the log level for the FOTA library.
+
+	  The available levels are:
+
+	  - 0 OFF: do not write
+	  - 1 ERROR: only write SYS_LOG_ERR
+	  - 2 WARNING: write SYS_LOG_WRN in addition to previous level
+	  - 3 INFO: write SYS_LOG_INF in addition to previous levels
+	  - 4 DEBUG: write SYS_LOG_DBG in addition to previous levels
+
 config FOTA_BLUEMIX_ORG
 	string "Bluemix organization code"
 	default "mstsez"
diff --git a/prj.conf b/prj.conf
index 828e847..5d6e50c 100644
--- a/prj.conf
+++ b/prj.conf
@@ -34,6 +34,7 @@ CONFIG_FOTA_DM_BACKEND_HAWKBIT=y
 # Logging
 CONFIG_SYS_LOG=y
 CONFIG_SYS_LOG_EXT_HOOK=y
+CONFIG_SYS_LOG_FOTA_LEVEL=4
 
 # Bluetooth (when enabled) debug
 CONFIG_BLUETOOTH_DEBUG_LOG=y
diff --git a/src/lib/bluemix.c b/src/lib/bluemix.c
index 7f7da3a..69fb4e2 100644
--- a/src/lib/bluemix.c
+++ b/src/lib/bluemix.c
@@ -10,7 +10,7 @@
  */
 
 #define SYS_LOG_DOMAIN "fota/bluemix"
-#define SYS_LOG_LEVEL SYS_LOG_LEVEL_DEBUG
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_FOTA_LEVEL
 #include <logging/sys_log.h>
 
 #include <zephyr.h>
diff --git a/src/lib/flash_block.c b/src/lib/flash_block.c
index f65d3bb..d329ec0 100644
--- a/src/lib/flash_block.c
+++ b/src/lib/flash_block.c
@@ -5,7 +5,7 @@
  */
 
 #define SYS_LOG_DOMAIN "fota/flash_block"
-#define SYS_LOG_LEVEL SYS_LOG_LEVEL_WARNING
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_FOTA_LEVEL
 #include <logging/sys_log.h>
 
 #include <stdint.h>
diff --git a/src/lib/hawkbit.c b/src/lib/hawkbit.c
index f6f5a59..265f83b 100644
--- a/src/lib/hawkbit.c
+++ b/src/lib/hawkbit.c
@@ -5,7 +5,7 @@
  */
 
 #define SYS_LOG_DOMAIN "fota/hawkbit"
-#define SYS_LOG_LEVEL SYS_LOG_LEVEL_DEBUG
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_FOTA_LEVEL
 #include <logging/sys_log.h>
 
 #include <stdint.h>
diff --git a/src/lib/mcuboot.c b/src/lib/mcuboot.c
index e18674c..b28e542 100644
--- a/src/lib/mcuboot.c
+++ b/src/lib/mcuboot.c
@@ -5,7 +5,7 @@
  */
 
 #define SYS_LOG_DOMAIN "fota/mcuboot"
-#define SYS_LOG_LEVEL SYS_LOG_LEVEL_DEBUG
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_FOTA_LEVEL
 #include <logging/sys_log.h>
 
 #include <stddef.h>
diff --git a/src/lib/tcp.c b/src/lib/tcp.c
index ee5705c..0e32c6c 100644
--- a/src/lib/tcp.c
+++ b/src/lib/tcp.c
@@ -5,7 +5,7 @@
  */
 
 #define SYS_LOG_DOMAIN "fota/tcp"
-#define SYS_LOG_LEVEL SYS_LOG_LEVEL_DEBUG
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_FOTA_LEVEL
 #include <logging/sys_log.h>
 
 #include <stdint.h>
```

I didn't use this option in the src/*.c files, because I want to move the Kconfig elsewhere.

**Comment by mike-scott** [comment]
@mbolivar I think src/bt_storage.c was missed for the CONFIG_SYS_LOG_FOTA_LEVEL change

**Comment by mike-scott** [comment]
Oh nevermind.. read your last comment "I didn't use this option in the src/*.c files, because I want to move the Kconfig elsewhere."

**Comment by rsalveti** [comment]
Looks fine for now, there is one minor checkpatch warning that can be fixed (WARNING:BLOCK_COMMENT_STYLE: Block comments use a trailing */ on a separate line), but happy to merge after that.

Would be nice to also be able to disable debug on src/main.c, but I'm assuming this will be in another patch-set.

**Comment by mbolivar** [comment]
Updated with additional checkpatch fix; diff from v3:

```
diff --git a/src/lib/mcuboot.c b/src/lib/mcuboot.c
index b28e542..7877aaa 100644
--- a/src/lib/mcuboot.c
+++ b/src/lib/mcuboot.c
@@ -77,8 +77,10 @@ uint8_t boot_status_read(void)
 void boot_status_update(void)
 {
        uint32_t offset;
-       /* The first byte of the Image OK area contains payload. The
-        * rest is padded with 0xff for flash write alignment. */
+       /*
+        * The first byte of the Image OK area contains payload. The
+        * rest is padded with 0xff for flash write alignment.
+        */
        uint8_t img_ok;
        uint8_t update_buf[TRAILER_IMAGE_OK_SIZE];
```

**Pull request merged**